File: Python/t24.py
# -*- conding:utf-8 -*-
#Author:lyc
import os, sys

# 题目：
# 给定一个链表，两两交换其中相邻的节点，并返回交换后的链表。
# 你不能只是单纯的改变节点内部的值，而是需要实际的进行节点交换。
#
# 示例:
# 给定 1->2->3->4, 你应该返回 2->1->4->3.

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        """
        设置三个指针移动
        :param head:
        :return:
        """
        # 链表为空或只有一个元素时返回
        if not head:return None
        if not head.next:return head

        result = ListNode(0)
        result.next = head

        first = result
        mid = head
        second = head.next

        while mid and second:
            # 交换节点
            mid.next = second.next
            second.next = mid
            first.next = second

            # 移动节点
            first = mid
            mid = first.next
            if mid:second=mid.next

        return result.next


    def showNode(self, node : ListNode) -> list:
        """
        show all value of ListNode
        :param node:
        :return:
        """
        nodeValue = []
        head = node
        while head:
            nodeValue.append(head.val)
            head = head.next
        return nodeValue


node = ListNode(1)
node.next = ListNode(2)
node.next.next = ListNode(3)
node.next.next.next = ListNode(4)

result = Solution()
new_node = result.swapPairs(node)
print(result.showNode(new_node))

File: Java/t24.java

/**
 * Definition for singly-linked list.
 */
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

/**
 * 利用三个指针遍历交换
 * Author:lyc
*/
class Solution {
    public ListNode swapPairs(ListNode head) {

        /**传入链表为空或者只有一个元素时直接返回*/
        if(head == null){
            return null;
        }

        if(head.next == null){
            return head;
        }

        /**创建带有保护节点的链表*/
        ListNode result = new ListNode(0);
        result.next = head;

        /**创建三个移动指针*/
        ListNode first = result;
        ListNode mid = head;
        ListNode second = head.next;

        while(mid != null && second != null)
        {
            // 交换节点
            mid.next = second.next;
            second.next = mid;
            first.next = second;

            // 移动节点
            first = mid;
            mid = first.next;
            if(mid != null){
                second = mid.next;
            }
        }
        return result.next;

    }

    /**
     * 打印显示链表的元素，用于结果验证
    */
    public void showListNode(ListNode node){
        while(node != null){
            System.out.print(node.val+"\t");
            node = node.next;
        }
    }
}


public class t24 {

    public static void main(String[] args){

        // 创建测试链表
        ListNode node = new ListNode(1);
        node.next = new ListNode(2);
        node.next.next = new ListNode(3);
        node.next.next.next = new ListNode(4);

        // 验证结果
        Solution result = new Solution();
        ListNode res = result.swapPairs(node);

        result.showListNode(res);
    }
    
}

